File: abbrase.js
//New Abbrase module
var abbrasePassphraseGen = function () {
    var unigrams, prefixes

    var ADJ = 0, NOUN = 1, VERB = 2

    function generate(length) {
        /* pick series of prefixes that will make up the passwords (Ryan Hitchman) */
        var numbers = new Uint16Array(length)
        getRandoms(numbers)

        //parts of speech: [adjectives] + noun + verb + [adjectives] + noun
        var noun = Math.floor((length - 3) / 2)
        var words = []

        for (var i = 0; i < length; i++) {
            var pre = prefixes[numbers[i] % 1024]
            var pos = ADJ
            if(i === noun || i === length - 1)
                pos = NOUN
            else if(i === noun + 1)
                pos = VERB
            words[i] = pre + getWordForPrefix(unigrams[pre], pos)
        }
        return words
    }

    function getRandoms(arr) {
        if (window.polyfilledCrypto) {
            //polyfill should overwrite Math.random()
            for (var i = 0; i < arr.length; i++) {
                arr[i] = Math.floor(Math.random() * 1024)
            }
        } else crypto.getRandomValues(arr)
    }

    function getWordForPrefix(prefix, pos) {
        if(prefix[pos].length)
            return choice(prefix[pos])
        else {
            //No words of the right type for this prefix :(
            var words = []
            for (var i = 0; i < prefix.length; i++) {
                for (var j = 0; j < prefix[i].length; j++) {
                    words.push(prefix[i][j])
                }
            }
            return choice(words)
        }
    }

    //pseudo-random as we're just choosing which suffix to use
    function choice(arr) {
        return arr[Math.floor(Math.random() * arr.length)]
    }

    return {
        generate: generate,
        load: function (json) {
            unigrams = json
            prefixes = Object.keys(unigrams)
        }
    }
}()
//End abbrase module

function show_phrase() {
    var length = Number($('#wordcount').val());
    if(!length || length < 3 || length > 10){
        length = 5
        $('#wordcount').val(5)
    }
    var words = abbrasePassphraseGen.generate(length),
        container = $('.current'),
        phrase = '', abbrase = ''
    for (var i = 0; i < length; i++) {
        var abbr = words[i].substring(0,3)
        var span = '<span>' + abbr + '</span>'
        phrase += span + words[i].substring(3) + ' '
        abbrase += span
    }

    var oldphrase = container.find('.phrase').html(),
        oldabbrase = container.find('.abbrase').html()
    $phrase = $('<div><span class="phrase">' + oldphrase + '</span></div>')
    $abbrase = $('<span class="abbrase">' + oldabbrase + '</span>').click(selectThis).prependTo($phrase)
    $('.history').prepend($phrase)

    container.find('.phrase').html(phrase)
    container.find('.abbrase').html(abbrase)
}

$.ajaxSetup({
    cache: true
})

function rngPolyfill() {
    //Check copied from Modernizr
    if ('crypto' in window && !!window.crypto && 'getRandomValues' in window.crypto)
        return true
    window.polyfilledCrypto  = true
    return $.getScript('//cdnjs.cloudflare.com/ajax/libs/seedrandom/2.3.6/seedrandom.min.js').then(function () {
        Math.seedrandom()
    })
}


var ready = $.Deferred()
$.when($.ajax({dataType: "json", url: 'unigrams.json'}), rngPolyfill(), ready)
    .done(function (response) {
        abbrasePassphraseGen.load(response[0])
        for (var i = 0; i < 12; i++) {
            show_phrase()
        }
        $('p.abbrase').attr('title', 'Your new password')
        $('.history div').last().prepend($('<div class="example">Example (do not use):</div>'))
        $('.abbrase').click(selectThis)
        $('#wordcount').change(show_phrase)
        $('.btn').click(function() {
            show_phrase()
            return false
        })
})

$(document).ready(function () {
    $('.btn').click(function () {return false})
    ready.resolve()
})

function selectThis() {
    //MDN
    var sel = window.getSelection()
    if (sel.isCollapsed) {
        var range = document.createRange()
        range.setStartBefore(this.firstChild)
        range.setEndAfter(this.lastChild)
        sel.removeAllRanges()
        sel.addRange(range)
    }
}

File: abbrase.py
"""abbrase.py: generates passwords by abbreviating passphrases"""

__author__ =    'Ben Caller (bcaller)'
__copyright__ = 'Copyright 2014, Benjamin Caller'
__credits__ =   ['Ryan Hitchman (rmmh)']
__license__ =   'MIT'

import json
import random
import math


def import_data():
    with open('unigrams.json') as unigrams:
        return json.load(unigrams)


data = import_data()
rand = random.SystemRandom()


def random_sample_with_replacement(seq, num):
    sequence = list(seq)
    return (sequence[rand.randrange(len(sequence))] for _ in range(num))


def phrase(words, pos):
    for prefix, i in zip(words, pos):
        if data[prefix][i]:
            yield prefix + random.choice(data[prefix][i])
        else:
            choices = [w for part in data[prefix] for w in part]
            if choices:
                yield random.choice(choices)
            else:
                yield prefix + '??'


def abbrase(abbrs=None, parts_of_speech=None, num=None, all_prefixes=set()):
    if parts_of_speech is None:
        zeroes = 2
        if not abbrs is None:  # default to: adj* noun verb adj* noun
            zeroes = len(abbrs) - 3
        elif not num is None:
            zeroes = num - 3
        parts_of_speech = [0] * math.floor(zeroes / 2) + [1, 2] + [0] * math.ceil(zeroes / 2) + [1]
    if abbrs is None:
        if len(all_prefixes) == 0:
            all_prefixes.update(set(data.keys()))

        abbrs = random_sample_with_replacement(all_prefixes, len(parts_of_speech))
    return ' '.join(phrase(abbrs, parts_of_speech))


def abbreviate(passphrase):
    return ''.join((w[:3] for w in passphrase.split(" ")))


def printable(passphrase):
    return abbreviate(passphrase) + ' | ' + passphrase

if __name__ == '__main__':
    for i in range(5):
        print(printable(abbrase(num=5)))

    print('\n')

    for i in range(2):
        print(printable(abbrase(num=6)))
    print('\n')
    for i in range(2):
        print(printable(abbrase(num=7)))